pkg/controller/job/job_controller.go

```diff
@@ -972,18 +1324,112 @@ func getBackoff(queue workqueue.RateLimitingInterface, key interface{}) time.Dur
 	return calculated
 }
 
-// countPodsByPhase returns pods based on their phase.
-func countPodsByPhase(job *batch.Job, pods []*v1.Pod, phase v1.PodPhase) int {
-	result := 0
+// countValidPodsWithFilter returns number of valid pods that pass the filter.
+// Pods are valid if they have a finalizer and, for Indexed Jobs, a valid
+// completion index.
+func countValidPodsWithFilter(job *batch.Job, pods []*v1.Pod, uncounted sets.String, filter func(*v1.Pod) bool) int {
+	result := len(uncounted)
 	for _, p := range pods {
-		idx := getCompletionIndex(p.Annotations)
-		if phase == p.Status.Phase && (!isIndexedJob(job) || (idx != unknownCompletionIndex && idx < int(*job.Spec.Completions))) {
+		// Pods that don't have a completion finalizer are in the uncounted set or
+		// have already been accounted for in the Job status.
+		if uncounted != nil && (!hasJobTrackingFinalizer(p) || uncounted.Has(string(p.UID))) {
+			continue
+		}
+		if isIndexedJob(job) {
+			idx := getCompletionIndex(p.Annotations)
+			if idx == unknownCompletionIndex || idx >= int(*job.Spec.Completions) {
+				continue
+			}
+		}
+		if filter(p) {
 			result++
 		}
 	}
 	return result
 }
 
+func trackingUncountedPods(job *batch.Job) bool {
+	return job.Status.UncountedTerminatedPods != nil && feature.DefaultFeatureGate.Enabled(features.JobTrackingWithFinalizers)
+}
+
+func appendJobCompletionFinalizerIfNotFound(finalizers []string) []string {
+	for _, fin := range finalizers {
+		if fin == batch.JobTrackingFinalizer {
+			return finalizers
+		}
+	}
+	return append(finalizers, batch.JobTrackingFinalizer)
+}
+
+func addTrackingFinalizerPatch(obj metav1.Object) []byte {
+	if hasJobTrackingFinalizer(obj) {
+		return nil
+	}
+	// Finalizers have patchStrategy=merge.
+	patch := map[string]interface{}{
+		"metadata": map[string]interface{}{
+			"finalizers": []string{batch.JobTrackingFinalizer},
+		},
+	}
+	patchBytes, _ := json.Marshal(patch)
+	return patchBytes
+}
+
+func removeTrackingFinalizerPatch(obj metav1.Object) []byte {
```

Comment:
Nit: unless I missed but I see this being executed only against pods, so you could change the signature to just pods. Same for `hasJobTrackingFinalizer`.

Reply:
It was for jobs and pods at some point. Reverted.